### PR TITLE
QRS-27, QRS-25, QRS-32 User and datasets management

### PR DESCRIPTION
**Pull request by VitalInnovator**
- added user tab to admin page
- Added the Guide page and a redirect to it
- Only users belonging to the organization have access to the dashboard. (member, admin)
- Guest is given to the user by default upon registration. Member is automatically assigned when the user is added to the organization
- Added created_at to the users table for convenient sorting
- Renamed Home page to Dashboard and updated folder structure
- Integrated a new panel on the admin page for dataset management
- Implemented functionalities to create new datasets by specifying their names
- Enabled administrators to set the status of datasets as 'active'
- A default limit of two active datasets was introduced; however, this limit's adjustability was made possible through an environment variable `NEXT_PUBLIC_MAX_ACTIVE_DATASETS` - Specifies the maximum number of active datasets allowed.
- Updated `getFragment` function, a new parameter called `datasetName` has been added. This parameter represents the name of the selected dataset. When making a request to the API endpoint, this parameter is included in the URL to specify the dataset from which to retrieve the fragment. 

The button is initially disabled and only becomes active if changes are made to the active datasets, enhancing user interaction
Enhanced the admin experience in the Organizations tab by restricting the capability to add datasets solely to active ones
Enhanced user satisfaction by ensuring that only active datasets are visible on their personalized dashboard pages

## Requested Python API changes
- Add a new parameter `datasetName` to the GET `/record` endpoint.
- Add a new POST endpoint `/datasets`  that accepts a list of active datasets `datasetNames` - an array of strings  `{datasetNames: [name1, name2]}`.

## Admin users

To change user into an admin you have to run this query:

```sql
UPDATE users SET role='admin'::users_role_enum WHERE email='user@email.com';
```

##  Instructions to Set Up the Application

1. Run the command `docker-compose build`.
2. Run the command `docker-compose up`.
3. Open `http://localhost:3000/` in your browser
4. Register a new user.
5. Use pgAdmin or another similar tool, and in the 'users' table, set the value of the 'role' field to 'admin'.
6. Refresh the page.
7. Navigate to the admin page.

   7.1.1. Go to the 'datasets' tab.

   7.1.2. Add a new dataset.
   
   7.1.3. Change the status of at least one dataset to 'active'.

   7.2.1. Go to the 'organizations' tab.

   7.2.2. Create a new organization and add users (including yourself) and one of the active datasets (IMPORTANT: only datasets marked as active are available for selection).

8. Navigate to the dashboard page.
9. Select one of the available datasets (IMPORTANT: only datasets marked as active are available for selection).
10. The application is ready to use.

## Important:
Due to caching, some users may experience a redirect to the guide page by default. To fix this unwanted redirect, you need to clear your browser's cache.

Screencast
https://drive.google.com/file/d/1Sbp1QMZWi81OEJW1PRo03C8roN0NZat6/view?usp=sharing

### CODE REVIEW

**Comment by roszcz** [comment]
@VitalInnovator thanks! Can you add some documentation in the readme about the user/orgranization/dataset management? Some kind of  a basic example about how to deal with the dataset access will be enough in the end

**Comment by roszcz** [comment]
@VitalInnovator Can you please make the documentation a priority? Everything looks great, but I'm not able to see any data without additional information on how to set up the dataset access.

image

**Comment by roszcz** [comment]
@VitalInnovator can you make instructions to actually get to a stage where I see signals on a dashboard explicit, please? I'm still unable to test this PR due to the lacking documentation.

File: next/src/modules/adminPage/components/SubTable.tsx [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        >
+          <Button
+            type="primary"
+            onClick={handleOpenCreateView}
+            style={{ marginBottom: 16 }}
+          >
+            {isMembershipType
+              ? 'Add New Membership'
+              : isDatasetAccessType
+              ? 'Add New Dataset'
```
Comment (roszcz) [inline next/src/modules/adminPage/components/SubTable.tsx]:
```suggestion
              ? 'Add Dataset Access'
```

I think this is more accurate

File: README.md [hunk 2]
```diff
@@ -1,10 +1,35 @@
...
 cp .env.example .env
 docker-compose build
 docker-compose up
 ```
 
+1. Run the command `docker-compose build`.
+2. Run the command `docker-compose up`.
+3. Open `http://localhost:3000/` in your browser
+4. Register a new user.
+5. Use pgAdmin or another similar tool, and in the 'users' table, set the value of the 'role' field to 'admin'.
```
Comment (roszcz) [inline README.md]:
```suggestion
5. Use pgAdmin or another similar tool, and in the 'users' table, set the value of the 'role' field to 'admin'. Raw query: `UPDATE users SET role='admin'::users_role_enum WHERE email='user@email.com';`
```

**Review by roszcz — CHANGES_REQUESTED** [review]
Thanks!

I have a couple change requests:

- [x] Change `datasetName` to `dataset_name` parameter name in the `/data` call
- [x] Change `datasetNames` to `dataset_names` in the `/datasets` call
- [x] In the Dataset Access table, when users click "Add Dataset Access", show all active and inactive datasets (now it shows only active)
- [x] Looks like deactivating and activating a dataset removes access? Access shouldn't go away when dataset is deactivated. (see video)
- [x] Show inactive datasets on the dashboard, add a note "Dataset inactive, concat admin to activate"

I left more directly on the code.

https://github.com/Nospoko/data-quality-platform/assets/8056825/1e3a9809-0440-493b-b1d1-829f7b4faa1b

**Comment by roszcz** [comment]
@VitalInnovator you can add an unused parameter in the python API in this PR:

```py
async def get_record(exam_uid: str, position: int, dataset_name: str):
  ...
```

And don't worry about the `/datasets` endpoint for now. When we close this, I'll work on the dataset sync and add this endpoint.

**Comment by VitalInnovator** [comment]
@roszcz I've implemented your feedback

**Comment by roszcz** [comment]
Here's some more cosmetic changes:

- [ ] If the user is logged in, point them to the dashboard, not to the guide.
- [x] In the datasets tab, change the buttons text from "Change State" to either "Activate" or "Deactivate", depending on the current dataset status.
- [x] In the request to `/datasets`, include the information about expected state, payload could look like this:
```json
{
  "datasets": [
    { "dataset_name": "foo", "state": "active" },
    { "dataset_name": "bar", "state": "inactive"},
  ]
}
```

**Review by roszcz — APPROVED** [review]

**Pull request merged**